libseccomp/src/error.rs

```diff
@@ -16,41 +32,59 @@ pub enum ErrorKind {
     Source,
 }
 
-#[derive(Debug)]
 pub struct SeccompError {
     kind: ErrorKind,
-    code: Option<i32>,
     source: Option<Box<dyn Error + Send + Sync>>,
 }
 
-impl fmt::Display for SeccompError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let msg = match &self.kind {
+impl SeccompError {
+    fn msg(&self) -> String {
+        match &self.kind {
             ErrorKind::Errno(e) => match -(*e) {
-                libc::EDOM => "Architecture specific failure",
-                libc::EACCES => "Setting the attribute with the given value is not allowed",
-                libc::EEXIST => "Failure regrading the existance of argument",
-                libc::EINVAL => "Invalid input",
-                libc::ENOMEM => "The library was unable to allocate enough memory",
-                libc::ECANCELED => "There was a system failure beyond the control of the library",
-                libc::EFAULT => "Internal libseccomp failure",
-                libc::ESRCH => "Unable to load the filter due to thread issues",
-                libc::EOPNOTSUPP => "The library doesn't support the particular operation",
-                errno => return write!(f, "Error caused by: errno({})", errno),
+                libc::EACCES => EACCES.to_string(),
+                libc::ECANCELED => ECANCELED.to_string(),
+                libc::EDOM => EDOM.to_string(),
+                libc::EEXIST => EEXIST.to_string(),
+                libc::EFAULT => EFAULT.to_string(),
+                libc::EINVAL => EINVAL.to_string(),
+                libc::ENOENT => ENOENT.to_string(),
+                libc::ENOMEM => ENOMEM.to_string(),
+                libc::EOPNOTSUPP => EOPNOTSUPP.to_string(),
+                libc::ERANGE => ERANGE.to_string(),
+                libc::ESRCH => ESRCH.to_string(),
+                errno => format!("Unknown error({})", errno),
             },
-            ErrorKind::Common(s) => s,
-            ErrorKind::ParseError => "Invalid argument",
-            ErrorKind::Source => return self.source.as_ref().unwrap().fmt(f),
-        };
-
-        if let Some(source) = &self.source {
-            write!(f, "{} caused by: {}", msg, source)
-        } else {
-            write!(f, "{}", msg)
+            ErrorKind::Common(s) => s.to_string(),
+            ErrorKind::ParseError => PARSE_ERROR.to_string(),
+            ErrorKind::Source => self.source.as_ref().unwrap().to_string(),
         }
     }
 }
 
+impl fmt::Display for SeccompError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let msg = self.msg();
+
+        if self.kind != ErrorKind::Source {
+            if let Some(source) = &self.source {
+                return write!(f, "{} caused by: {}", msg, source);
+            }
+        }
+
+        write!(f, "{}", msg)
```

Comment:
Is this more readable? Or do we have to live with the fact that this can not expressed readable without `&& if let`?
```suggestion
        match &self.source {
            Some(source) if !matches!(self.kind, ErrorKind::Source) => {
                write!(f, "{} caused by: {}", msg, source);
            }
            Some(_source) | None => {
                write!(f, "{}", msg)
            }
        }
```

Reply:
I fixed it. Thanks.